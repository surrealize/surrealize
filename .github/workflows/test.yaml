name: Test & Build

on:
  push:
    branches:
      - "**"
      - "!next"
      - "!main"
  workflow_call:

jobs:
  build:
    name: Test & Build
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: oven-sh/setup-bun@v2
        with:
          bun-version: latest
      - uses: surrealdb/setup-surreal@v2
        with:
          surrealdb_auth: false
      - run: bun install --frozen-lockfile
      - run: bun run check
      - run: bun test
      - run: bun run build
