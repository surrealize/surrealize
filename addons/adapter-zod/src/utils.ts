import type { z } from "zod";

/**
 * Shorthand for creating a Zod custom type.
 */
export type ZodCustom<T> = z.ZodType<T, z.ZodTypeDef, T>;
