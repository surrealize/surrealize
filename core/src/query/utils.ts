import { RawQuery } from "./builder/raw.ts";
import type { Query } from "./query.ts";
import { type TaggedTemplate, format, merge, tagString } from "./template.ts";
import { type PreparedQuery, type QueryLike } from "./types.ts";

export const resolveQuery = <TOutput>(
	query: QueryLike<TOutput>,
): Query<TOutput> => {
	if ("toQuery" in query) {
		const toQueryValue = query.toQuery();

		if (toQueryValue instanceof RawQuery) {
			return toQueryValue.toQuery();
		}

		return toQueryValue;
	}

	return query;
};

/**
 * Convert a tagged template query to a prepared query by replacing variables with bindings.
 *
 * Also encode the variables to their surrealdb.js representation if possible.
 *
 * @param query The tagged template to convert.
 * @returns The compiled query (including the query string and the bindings object).
 */
export const prepareQuery = (template: TaggedTemplate): PreparedQuery => {
	const normalizedQuery = stripLastSemicolon(template);
	const formattedQuery = format(normalizedQuery, (_, index) => `$_${index}`);
	const bindings = Object.fromEntries(
		normalizedQuery[1].map((value, index) => [`_${index}`, value] as const),
	);

	return { query: formattedQuery, bindings };
};

/**
 * Convert a tagged template query list to a compiled query by replacing variables with bindings and wrapping it in a transaction.
 *
 * Also encode the variables to their surrealdb.js representation if possible.
 *
 * @param queries The tagged template queries to convert and wrap in a transaction.
 * @returns The compiled query (including the query string and the bindings object) as a transaction.
 */
export const prepareTransaction = (queries: Query[]): PreparedQuery => {
	return prepareQuery(
		merge([
			merge(
				// wrap the queries in a transaction
				[
					tagString("BEGIN"),
					...queries.map((query) =>
						merge([tagString("("), query.template, tagString(")")]),
					),
					tagString("COMMIT"),
				],
				";\n",
			),
			tagString(";"),
		]),
	);
};

/**
 * Strip the last semicolon from a tagged template query.
 *
 * @param query The tagged template query to strip the last semicolon from.
 * @returns The tagged template query without an ending semicolon.
 */
const stripLastSemicolon = (query: TaggedTemplate): TaggedTemplate => {
	return [
		query[0].map((sequence, index, array) => {
			// check if it is the last sequence and ends with a semicolon
			if (index === array.length - 1 && sequence.endsWith(";")) {
				return sequence.slice(0, -1);
			}

			return sequence;
		}),
		query[1],
	];
};
