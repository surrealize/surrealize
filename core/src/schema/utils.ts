import type { StandardSchemaV1 } from "@standard-schema/spec";

import {
	type InferSchemaOutput,
	type Schema,
	ValidationError,
} from "./types.ts";

/**
 * Validates a value against a schema and returns the validated value.
 *
 * If the schema validation fails, an error will be thrown.
 *
 * @param schema The schema to validate against.
 * @param value The value to validate.
 * @returns The validated value or an error if the validation failed.
 */
export const parseSchema = async <TSchema extends Schema>(
	schema: TSchema,
	value: unknown,
): Promise<InferSchemaOutput<TSchema>> => {
	const standard = schema["~standard"];

	let result = standard.validate(value);
	if (result instanceof Promise) result = await result;

	if (result.issues) throw new ValidationError(result.issues);

	return result.value as InferSchemaOutput<TSchema>;
};

export const mergeSchema = <TInput, TOutput>(
	schemas: Schema<TInput, TOutput>[],
): Schema<TInput, TOutput> => {
	const issues: StandardSchemaV1.Issue[] = [];

	const validate = async (
		value: unknown,
	): Promise<StandardSchemaV1.Result<TOutput>> => {
		for (const schema of schemas) {
			try {
				const parsedValue = await parseSchema(schema, value);
				return { value: parsedValue };
			} catch (error) {
				if (error instanceof ValidationError) {
					issues.push(...error.issues);
					// if the schema fails, try the next one
				} else {
					// throw if the error in unrecognized
					throw error;
				}
			}
		}

		return { issues };
	};

	return {
		["~standard"]: {
			version: 1,
			vendor: "surrealize",
			validate,
		},
	};
};
