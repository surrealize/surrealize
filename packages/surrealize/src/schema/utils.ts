import type { SchemaLike } from "./types.ts";

/**
 * Validates a value against a schema and returns the validated value.
 *
 * If the schema validation fails, an error will be thrown.
 *
 * @param schema The schema to validate against.
 * @param value The value to validate.
 * @returns The validated value or an error if the validation failed.
 */
export const parseSchema = <TSchemaOutput>(
	schema: SchemaLike<TSchemaOutput>,
	value: unknown,
): TSchemaOutput => {
	switch (true) {
		case typeof schema === "function":
			return schema(value);
		case typeof schema === "object":
			// like zod
			if ("parse" in schema) return schema.parse(value);
			// like yup
			if ("validate" in schema) return schema.validate(value);
	}
	throw new Error("Invalid schema");
};

export const mergeSchema = <TSchemaOutput>(
	schemas: SchemaLike<TSchemaOutput>[],
): SchemaLike<TSchemaOutput> => {
	const errors: unknown[] = [];

	// create a new schema function which will try each schema in order
	return (value) => {
		for (const schema of schemas) {
			// try each schema
			try {
				return parseSchema(schema, value);
			} catch (error) {
				errors.push(error);
				// if the schema fails, try the next one
			}
		}

		// if all schemas fail, throw an error
		throw new Error(
			"All schemas failed. Expected at least one schema to pass.",
			{ cause: errors },
		);
	};
};
