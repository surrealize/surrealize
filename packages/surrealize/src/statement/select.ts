import {
	type Builder,
	createBuilder,
	createStatement,
} from "../query/builder/statements.ts";
import { merge, tag, tagString } from "../query/template.ts";
import {
	type InferFields,
	enforceField,
	enforceFields,
} from "../query/validation/field.ts";
import { type TargetLike, resolveTarget } from "../type/target.ts";
import { type WhereCondition, buildWhere } from "./shared/where.ts";

export const select = <TSchema>() =>
	createStatement((raw, ctx) => {
		return (
			...fields: InferFields<TSchema>[]
		): Builder<{
			from: typeof from<TSchema>;
			fromOnly: typeof fromOnly<TSchema>;
		}> => {
			const appended = raw.append(
				tagString(
					`SELECT ${fields.length === 0 ? "*" : enforceFields(fields, "wildcard").join(", ")}`,
				),
			);

			return createBuilder(appended, ctx, { from, fromOnly });
		};
	});

export const selectValue = <TSchema>() =>
	createStatement((raw, ctx) => {
		return (
			field: InferFields<TSchema>,
		): Builder<{
			from: typeof from<TSchema>;
			fromOnly: typeof fromOnly<TSchema>;
		}> => {
			const appended = raw.append(
				tagString(`SELECT VALUE ${enforceField(field, "wildcard")}`),
			);

			return createBuilder(appended, ctx, { from, fromOnly });
		};
	});

const from = <TSchema>() =>
	createStatement(
		(raw, ctx) =>
			(
				...targets: TargetLike[]
			): Builder<{
				parallel: typeof parallel<TSchema>;
			}> =>
				createBuilder(
					raw.append(
						merge(
							[
								tagString("FROM"),
								merge(
									targets.map((target) => tag`${resolveTarget(target)}`),
									", ",
								),
							],
							" ",
						),
					),
					ctx,
					{ parallel },
				),
	);

const fromOnly = <TSchema>() =>
	createStatement(
		(raw, ctx) =>
			(
				target: TargetLike,
			): Builder<{
				where: typeof where<TSchema>;
				parallel: typeof parallel<TSchema>;
			}> =>
				createBuilder(
					raw.append(tag`FROM ONLY ${resolveTarget(target)}`),
					ctx,
					{
						where,
						parallel,
					},
				),
	);

const where = <TSchema>() =>
	createStatement(
		(raw, ctx) =>
			(
				...conditions: WhereCondition<TSchema>[]
			): Builder<{
				parallel: typeof parallel<TSchema>;
			}> => {
				return createBuilder(raw.append(buildWhere(conditions)), ctx, {
					parallel,
				});
			},
	);

const parallel = <TSchema>() =>
	createStatement(
		(raw, ctx) => () =>
			createBuilder(raw.append(tagString("PARALLEL")), ctx, {}),
	);
