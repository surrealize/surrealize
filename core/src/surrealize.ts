import type { EventEmitter } from "./connection/emitter.ts";
import {
	AbstractEngine,
	type EmitterEvents,
	type EngineInitializer,
} from "./connection/engine.ts";
import { WebSocketEngine } from "./connection/engines/ws.ts";
import { DatabaseError, QueryError } from "./connection/error.ts";
import type { RpcRequest, RpcResponse } from "./connection/rpc.ts";
import type { Auth } from "./connection/types.ts";
import { surql } from "./query/query.ts";
import type {
	InferQueriesOutput,
	QueriesLike,
	QueryLike,
} from "./query/types.ts";
import {
	prepareQuery,
	prepareTransaction,
	resolveQuery,
} from "./query/utils.ts";
import type { Schema } from "./schema/types.ts";
import { parseSchema } from "./schema/utils.ts";
import { RecordId } from "./type/recordid.ts";
import { type TargetLike, resolveTarget } from "./type/target.ts";

export type SurrealizeOptions = {
	url: URL | string;

	namespace?: string;
	database?: string;
	auth?: Auth;

	/**
	 * Set the connection as the default connection.
	 *
	 * This will make the Surrealize instance the default instance for all queries.
	 */
	default?: boolean;

	/**
	 * A collection of engines which are available to connect to.
	 *
	 * Defaults to the following engines:
	 * - `ws`: WebSocket
	 * - `wss`: WebSocket Secure
	 */
	engines?: Record<string, EngineInitializer>;

	/**
	 * The timeout in milliseconds for the connection.
	 */
	timeout?: number;
};

export class Surrealize {
	static default: Surrealize | undefined = undefined;

	readonly #engine: AbstractEngine;
	readonly emitter: EventEmitter<EmitterEvents>;

	constructor(options: SurrealizeOptions) {
		const engines: Record<string, EngineInitializer> = options.engines ?? {
			ws: (ctx) => new WebSocketEngine(ctx),
			wss: (ctx) => new WebSocketEngine(ctx),
		};
		const url = options.url instanceof URL ? options.url : new URL(options.url);

		const protocol = url.protocol.replace(":", "");
		const engine = engines[protocol];

		if (!engine) throw new Error(`Unsupported protocol ${protocol}`);

		this.#engine = engine({
			url,
			namespace: options.namespace,
			database: options.database,
			auth: options.auth,
			timeout: options.timeout,
		});
		this.emitter = this.#engine.emitter;

		if (options.default) Surrealize.default = this;
	}

	async connect(): Promise<void> {
		return this.#engine.connect();
	}

	async disconnect(): Promise<void> {
		return this.#engine.disconnect();
	}

	async version(): Promise<string> {
		return this.#engine.version();
	}

	async execute<TSchemaOutput>(
		queryLike: QueryLike<TSchemaOutput>,
	): Promise<TSchemaOutput> {
		const { template, schema } = resolveQuery(queryLike);
		const { query, bindings } = prepareQuery(template);

		const [result] = await this.query(query, bindings);

		return (schema ? parseSchema(schema, result) : result) as TSchemaOutput;
	}

	async executeAll<const TQueries extends QueriesLike>(
		queriesLike: TQueries,
	): Promise<InferQueriesOutput<TQueries>> {
		return Promise.all(
			queriesLike.map((queryLike) =>
				this.execute(queryLike),
			) as InferQueriesOutput<TQueries>,
		);
	}

	async executeTransaction<const TQueries extends QueriesLike>(
		queriesLike: TQueries,
	): Promise<InferQueriesOutput<TQueries>> {
		const queries = queriesLike.map((queryLike) => resolveQuery(queryLike));

		const { query, bindings } = prepareTransaction(queries);

		const results = await this.query(query, bindings);

		return queries.map(({ schema }, index) => {
			const queryResult = results[index];
			return schema ? parseSchema(schema, queryResult) : queryResult;
		}) as InferQueriesOutput<TQueries>;
	}

	/**
	 * Get a specific target from the database. A target can be a record id or a table.
	 * In case the target is a table, the result will be an array of records.
	 *
	 * @param targetLike A target like input (record id or table).
	 * @param schema The optional schema to use for validating the result.
	 * @returns The target from the database.
	 */
	async resolve<TOutput>(
		targetLike: TargetLike,
		schema?: Schema<TOutput>,
	): Promise<TOutput> {
		const target = resolveTarget(targetLike);

		const query =
			target instanceof RecordId
				? surql`SELECT * FROM ONLY ${target}`
				: surql`SELECT * FROM ${target}`;

		return this.execute(query.withSchema(schema));
	}

	async rpc<TResult>(request: RpcRequest): Promise<RpcResponse<TResult>> {
		return this.#engine.rpc(request);
	}

	async query<TResult extends unknown[]>(
		query: string,
		bindings?: Record<string, unknown>,
	): Promise<TResult> {
		const response = await this.rpc<
			Array<{ result: unknown; status: "OK" | "ERR"; time: string }>
		>({
			method: "query",
			params: [query, bindings],
		});

		if (response.result) {
			const error = response.result.find((query) => query.status === "ERR");
			if (error) throw new QueryError(error.result as string);

			return response.result.map((query) => query.result) as TResult;
		} else {
			throw new DatabaseError(response.error);
		}
	}
}
