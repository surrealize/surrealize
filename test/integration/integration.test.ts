import { afterEach, beforeAll, describe, expect, test } from "bun:test";
import { Surrealize, WebSocketEngine, surql } from "surrealize";

import { testSelectOnly, testSelectWhere } from "./jobs/select.ts";
import { testVersion } from "./jobs/version.ts";
import { cleanupDemoData } from "./utils.ts";

const surrealize: Surrealize = new Surrealize(
	new WebSocketEngine("ws://localhost:8000", {
		namespace: "test",
		database: "test",
	}),
);

beforeAll(async () => {
	await surrealize.connect();
});

afterEach(async () => {
	await cleanupDemoData(surrealize);
});

describe("Integration", () => {
	test("connection", async () => {
		const random = Math.floor(Math.random() * 100);
		const result = await surrealize.execute(surql`RETURN ${random}`);

		expect(result).toEqual(random);
	});

	test("version", () => testVersion(surrealize));

	test("select from where", () => testSelectWhere(surrealize));

	test("select from only", () => testSelectOnly(surrealize));

	test("None and Null", async () => {
		expect(await surrealize.execute(surql`RETURN NONE`)).toEqual(undefined);
		expect(await surrealize.execute(surql`RETURN NULL`)).toEqual(null);
	});
});
