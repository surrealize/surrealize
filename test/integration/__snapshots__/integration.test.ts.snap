// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration filter 1`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 10,
    "id": RecordId {
      "table": "user",
      "value": "bob",
    },
    "name": "Bob",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration filter 2`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
]
`;

exports[`Integration filter 3`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from where 1`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 10,
    "id": RecordId {
      "table": "user",
      "value": "bob",
    },
    "name": "Bob",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from where 2`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
]
`;

exports[`Integration select from where 3`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from only 1`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 10,
    "id": RecordId {
      "table": "user",
      "value": "bob",
    },
    "name": "Bob",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from only 2`] = `
{
  "age": 10,
  "id": RecordId {
    "table": "user",
    "value": "bob",
  },
  "name": "Bob",
}
`;

exports[`Integration select from where 4`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 10,
    "id": RecordId {
      "table": "user",
      "value": "bob",
    },
    "name": "Bob",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from where 5`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
]
`;

exports[`Integration select from where 6`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from only 3`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 10,
    "id": RecordId {
      "table": "user",
      "value": "bob",
    },
    "name": "Bob",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from only 4`] = `
{
  "age": 10,
  "id": RecordId {
    "table": "user",
    "value": "bob",
  },
  "name": "Bob",
}
`;

exports[`Integration select from where 7`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 10,
    "id": RecordId {
      "table": "user",
      "value": "bob",
    },
    "name": "Bob",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from where 8`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
]
`;

exports[`Integration select from where 9`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from only 5`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 10,
    "id": RecordId {
      "table": "user",
      "value": "bob",
    },
    "name": "Bob",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from only 6`] = `
{
  "age": 10,
  "id": RecordId {
    "table": "user",
    "value": "bob",
  },
  "name": "Bob",
}
`;

exports[`Integration select from where 10`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 10,
    "id": RecordId {
      "table": "user",
      "value": "bob",
    },
    "name": "Bob",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from where 11`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
]
`;

exports[`Integration select from where 12`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from only 7`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 10,
    "id": RecordId {
      "table": "user",
      "value": "bob",
    },
    "name": "Bob",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from only 8`] = `
{
  "age": 10,
  "id": RecordId {
    "table": "user",
    "value": "bob",
  },
  "name": "Bob",
}
`;

exports[`Integration select from where 13`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 10,
    "id": RecordId {
      "table": "user",
      "value": "bob",
    },
    "name": "Bob",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from where 14`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
]
`;

exports[`Integration select from where 15`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from only 9`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 10,
    "id": RecordId {
      "table": "user",
      "value": "bob",
    },
    "name": "Bob",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from only 10`] = `
{
  "age": 10,
  "id": RecordId {
    "table": "user",
    "value": "bob",
  },
  "name": "Bob",
}
`;

exports[`Integration select from where 16`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 10,
    "id": RecordId {
      "table": "user",
      "value": "bob",
    },
    "name": "Bob",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from where 17`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
]
`;

exports[`Integration select from where 18`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from only 11`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 10,
    "id": RecordId {
      "table": "user",
      "value": "bob",
    },
    "name": "Bob",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from only 12`] = `
{
  "age": 10,
  "id": RecordId {
    "table": "user",
    "value": "bob",
  },
  "name": "Bob",
}
`;

exports[`Integration select from where 19`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 10,
    "id": RecordId {
      "table": "user",
      "value": "bob",
    },
    "name": "Bob",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from where 20`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
]
`;

exports[`Integration select from where 21`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from only 13`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 10,
    "id": RecordId {
      "table": "user",
      "value": "bob",
    },
    "name": "Bob",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from only 14`] = `
{
  "age": 10,
  "id": RecordId {
    "table": "user",
    "value": "bob",
  },
  "name": "Bob",
}
`;

exports[`Integration select from where 22`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 10,
    "id": RecordId {
      "table": "user",
      "value": "bob",
    },
    "name": "Bob",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from where 23`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
]
`;

exports[`Integration select from where 24`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from only 15`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 10,
    "id": RecordId {
      "table": "user",
      "value": "bob",
    },
    "name": "Bob",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from only 16`] = `
{
  "age": 10,
  "id": RecordId {
    "table": "user",
    "value": "bob",
  },
  "name": "Bob",
}
`;

exports[`Integration select from where 25`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 10,
    "id": RecordId {
      "table": "user",
      "value": "bob",
    },
    "name": "Bob",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from where 26`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
]
`;

exports[`Integration select from where 27`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from only 17`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 10,
    "id": RecordId {
      "table": "user",
      "value": "bob",
    },
    "name": "Bob",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration select from only 18`] = `
{
  "age": 10,
  "id": RecordId {
    "table": "user",
    "value": "bob",
  },
  "name": "Bob",
}
`;
